### PR TITLE
Fix constant reference check in CFA

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR is a follow up fix to #44730. The PR implements a more thorough check to ensure references are non-mutable when they are narrowed by aliased conditional expressions. For example:

```ts
function test(obj: { readonly x: string | number }) {
    const isString = typeof obj.x === 'string';
    obj = { x: 42 };
    if (isString) {
        let s: string = obj.x;  // Not narrowed because obj is assigned in function body
    }
}
```

Previously we'd just check that `x` is `readonly`. Now we also check that `obj` isn't assigned in the function body.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot perf test this

**Comment by ahejlsberg** [comment]
The slight performance improvement is likely from deferring more calls to `isParameterAssigned` (which can result in an expensive tree walk).

@RyanCavanaugh Would be nice it we could cherry pick this fix into the 4.4 beta.

**Comment by DanielRosenwasser** [comment]
If we merge this in, I'll sync the release-4.4 branch and rebuild later this afternoon with @uniqueiniquity since we have 1-2 other PRs that we want in too.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22975,7 +22975,20 @@ namespace ts {
...
 
-        function getFlowTypeOfReference(reference: Node, declaredType: Type, initialType = declaredType, isConstant?: boolean, flowContainer?: Node) {
+        function isConstantReference(node: Node): boolean {
+            switch (node.kind) {
+                case SyntaxKind.Identifier:
+                    const symbol = getResolvedSymbol(node as Identifier);
+                    return isConstVariable(symbol) || !!symbol.valueDeclaration && getRootDeclaration(symbol.valueDeclaration).kind === SyntaxKind.Parameter && !isParameterAssigned(symbol);
+                case SyntaxKind.PropertyAccessExpression:
+                case SyntaxKind.ElementAccessExpression:
+                    // The resolvedSymbol property is initialized by checkPropertyAccess or checkElementAccess before we get here.
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:22985]: Would be nice if there was a way to assert this easily.

File: tests/cases/conformance/controlFlow/controlFlowAliasing.ts [hunk 2]
```diff
@@ -48,6 +48,36 @@ function f14(x: number | null | undefined): number | null {
...
+}
+
+function f18(obj: readonly [string | number]) {
+    const isString = typeof obj[0] === 'string';
+    obj = [42];
+    if (isString) {
+        let s: string = obj[0];  // Not narrowed because of is assigned in function body
+    }
+}
+
```
Comment (DanielRosenwasser) [inline tests/cases/conformance/controlFlow/controlFlowAliasing.ts:80]: Maybe a worthwhile test - a parameter property that is marked as `readonly` and which is not assigned in its body. 

```suggestion

class C {
    constructor(readonly x: string | number) {
        const thisX_isString = typeof this.x === 'string';
        const xIsString = typeof x === 'string';
        if (thisX_isString && xIsString) {
            let s: string;
            s = obj.x;
            s = x;
        }
    }
}
```

**Comment by DanielRosenwasser** [comment]
I added the tests case I mentioned in https://github.com/microsoft/TypeScript/pull/44762/commits/3b12c042546aea831c2a2b8df096a7bca31df552

Kind of surprised we don't narrow `this.x`, but not worth blocking on.

**Pull request merged**